subsys/logging/log_backend_adsp_mtrace.c

```diff
@@ -0,0 +1,221 @@
+
+#define MTRACE_CORE		0
+
+struct adsp_debug_slot {
+	uint32_t host_ptr;
+	uint32_t dsp_ptr;
+	uint8_t data[ADSP_DW_SLOT_SIZE - sizeof(uint32_t) * 2];
+} __packed;
+
+static void mtrace_init(void)
+{
+	if (ADSP_DW->descs[0].type == MTRACE_LOGGING_SLOT_TYPE(MTRACE_CORE)) {
+		return;
+	}
+
+	ADSP_DW->descs[0].type = MTRACE_LOGGING_SLOT_TYPE(MTRACE_CORE);
+}
+
+static void mtrace_out(int8_t *str, size_t len, size_t *space_left)
+{
+	struct adsp_debug_slot *slot = (struct adsp_debug_slot *)(ADSP_DW->slots[0]);
+	uint8_t *data = slot->data;
+	uint32_t r = slot->host_ptr;
+	uint32_t w = slot->dsp_ptr;
+	size_t avail, left;
+
+	if (w > r) {
+		avail = MTRACE_LOG_BUF_SIZE - w + r - 1;
+	} else if (w == r) {
+		avail = MTRACE_LOG_BUF_SIZE - 1;
+	} else {
+		avail = r - w - 1;
+	}
+
+	if (len == 0) {
+		goto out;
+	}
+
+	/* data that does not fit is dropped */
+	left = MIN(avail, len);
+
+	if (w + left >= MTRACE_LOG_BUF_SIZE) {
+		size_t tail = MTRACE_LOG_BUF_SIZE - w;
+
+		memcpy(data + w, str, tail);
+		left -= tail;
+		memcpy(data, str + tail, left);
+		w = left;
+	} else {
+		memcpy(data + w, str, left);
+		w += left;
+	}
+
+	slot->dsp_ptr = w;
+
+out:
+	if (space_left) {
+		*space_left = avail > len ? avail - len : 0;
+	}
+}
+
+static int char_out(uint8_t *data, size_t length, void *ctx)
+{
+	size_t space_left = 0;
+	/*
+	 * we handle the data even if mtrace is active. this
+	 * ensures we can capture early boot messages.
+	 */
+	mtrace_out(data, length, &space_left);
+
+	if (mtrace_active && mtrace_hook) {
+		mtrace_hook(length, space_left);
+	}
+
+	return length;
+}
+
+/**
+ * 80 bytes seems to catch most sensibly sized log message lines
+ * in one go letting the trace out call output whole complete
+ * lines. This avoids the overhead of a spin lock in the trace_out
+ * more often as well as avoiding entwined characters from printk if
+ * LOG_PRINTK=n.
+ */
+#define LOG_BUF_SIZE 80
+static uint8_t log_buf[LOG_BUF_SIZE];
+
+LOG_OUTPUT_DEFINE(log_output_adsp_mtrace, char_out, log_buf, sizeof(log_buf));
+
+static uint32_t format_flags(void)
+{
+	uint32_t flags = LOG_OUTPUT_FLAG_LEVEL | LOG_OUTPUT_FLAG_TIMESTAMP;
+
+	if (IS_ENABLED(CONFIG_LOG_BACKEND_FORMAT_TIMESTAMP)) {
+		flags |= LOG_OUTPUT_FLAG_FORMAT_TIMESTAMP;
+	}
+
+	return flags;
+}
+
+static void panic(struct log_backend const *const backend)
+{
+	k_spinlock_key_t key = k_spin_lock(&mtrace_lock);
+
+	log_backend_std_panic(&log_output_adsp_mtrace);
+
+	k_spin_unlock(&mtrace_lock, key);
+}
+
+static void dropped(const struct log_backend *const backend,
+		    uint32_t cnt)
+{
+	log_output_dropped_process(&log_output_adsp_mtrace, cnt);
+}
+
+static void process(const struct log_backend *const backend,
+		    union log_msg_generic *msg)
+{
+	log_format_func_t log_output_func = log_format_func_t_get(log_format_current);
+
+	k_spinlock_key_t key = k_spin_lock(&mtrace_lock);
+
+	log_output_func(&log_output_adsp_mtrace, &msg->log, format_flags());
```

Comment:
imo, this approach can lead to issues. In deferred mode you are locking in the lowest priority thread for time that is unknown since it depends on the log message string. You can have here short string with no arguments or string with 12 sophisticated arguments (`%-20.5s`).
I'm not sure about architecture but does lock work for exception (memfault, etc.) which leads to panic?

Assuming that lock prevents preemption i was thinking about following approach: 
- in process (if not in panic): store pointer to the log message and format message string to a ram buffer. When formatting is done, under lock dump the buffer and reset message pointer.
- in panic, check if pointer to the current log message is set, if yes it means that process was interrupted (but data is not yet dumped). Process this message and set the flag to directly process messages in the process function.

Reply:
Thanks @nordic-krch for the note! This is same approach as already used in subsys/logging/log_backend_adsp.c , added in commit 02503c37c3c983a089b2ab69ad1a5f58024cdda3 ("logging: backend_adsp: lock the output") by @teburd . Let's loop him in here.

I'm not super familiar with how Zephyr handles the panics (and in which conditions the log backend panic may be called), but the scenario you describe seems possible. If this holds, even a small time window to keep the lock in process may be problematic and maybe we just have to do a try_lock in panic(). If the time window is hit, panic log cannot be written, but at least we don't get stuck spinning forever on the lock.

Reply:
@nordic-krch @teburd I now simplified the implementation and do not call the log_backend_std_panic() at all in panic. It seems we have no "try_lock" variant for spinlocks in Zephyr. Hmm. And that wouldn't be super clean either. As I couldn't figure out a non-convoluted way to make this truly safe, and based on testing with actual panics on hw, we don't rellly need to do the std_panic() here. So alas, panic simplified in latest push.

Reply:
I'm not sure this solves the problem @nordic-krch mentioned yet as the format is still being done inside the spin lock

I believe what @nordic-krch is suggesting (I could be mis-interpretting) is something like...

```c
char temp_fmt[SOME_BUF]; /* context local buffer for writting formatted log message */
log_output_func(...);
k_spinlock_key_t key = k_spin_lock(&lock);
mtrace_out(temp_fmt, ..);
k_spin_unlock(&lock, key);
```

Such that the formatting is done outside the spin lock.

The spin lock was/is there to prevent a race between one core in an exception and another running the log process thread where both things may be attempting to write, in different context, to the backend.

Reply:
yes, @teburd that was my idea, to take string formatting outside of the lock.

Reply:
Hmm, @teburd and @nordic-krch , but that just shortens the window of a possible race, right? If  exception during formatting is a problem, same can happen during mtrace_out().

Maybe this is still a prudent thing to, but I now wonder if this is really needed.

Let's say we have a system with two cores. If coreA is (slowly) formatting a log entry (with lock taken) and coreB hits an exception. With my latest patch revision, core B will go to z_impl_log_panic and do a "while (log_process() == true) {}" loop. This will end to mtrace backends process() callback and coreB will spin on the lock until A is ready. But this is what we desire. The system is a lost case, so why not write the logs out. If the panic happens in formatting function of coreA, we are screwed anyway, and in the system is in a state where it cannot execute z_impl_log_panic() anymore, we are unlikely to get sensible logs out from the formatters anyway.

So would seem the current approach is still fairly reasonable.

Reply:
please be aware of the fact that i'm not that competent when it comes to multicore/smp environment. Does spin_lock locks all interrupts on given core and prevents preemption (it maps to irq_lock on single core cpu's like cortex-m)? if yes then my concern was that locking around string formatting means that higher priority tasks and interrupts cannot preempt lowest priority task and that can randomly cause issues.

Reply:
(That's exactly how spinlocks work, yes)